tests/p4ctest.h

```diff
@@ -0,0 +1,165 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef P4CTEST_H
+#define P4CTEST_H
+
+#include <iostream>
+#include <unordered_map>
+#include <vector>
+#include <string>
+#include <cassert>
+#include "lib/stringify.h"
+
+#define ASSERT_EQ(a, b) \
+    do {                  \
+        if ((a) != (b)) {                                               \
+            std::clog << cstring("") << #a << " is `" << Util::toString(a) << "'" << std::endl \
+                      << cstring("") << #b << " is `" << Util::toString(b) << "'" << std::endl; \
+            assert((a) == (b));                                         \
+        }                                                               \
+    } while (0)
+
+#define ASSERT_NEQ(a, b) \
+    do {                  \
+        if ((a) == (b)) {                                               \
+            std::clog << cstring("") << #a << " is `" << Util::toString(a) << "'" << std::endl \
+                      << cstring("") << #b << " is `" << Util::toString(b) << "'" << std::endl; \
+            assert((a) != (b));                                         \
+        }                                                               \
+    } while (0)
+
+#define UNREACHABLE()                           \
+    do {                                                   \
+        std::clog << "Should not reach here" << std::endl; \
+        assert(false);                                     \
+    } while (0)
+
+#define RUNTEST(f)  do {  \
+        int result = f();                     \
+        if (result != SUCCESS) return result; \
+    } while (0)
+
+namespace Test {
+class TestBase {
+ public:
+    virtual ~TestBase() {}
+    virtual int run() = 0;
+
+    const int SUCCESS = 0;
+    const int FAILURE = 1;
+};
+
+struct TestCommandContext {
+    /* number of command line arguments */
+    int argc;
+    /* array of command line arguments */
+    char **argv;
+    /* private context data defined by the API user */
+    void *pvt;
+};
+
+typedef std::function<void(TestCommandContext* ctx)> CommandHandler;
+
+class TestCommand {
+ public:
+    TestCommand(std::string name, CommandHandler handler) : name(name), handler(handler) {}
+    CommandHandler handler;
+    std::string name;
+};
+
+typedef std::vector<TestCommand> TestCommands;
+
+class TestDriver {
+ static TestDriver singleton;
+ public:
+    static TestCommands commands;
+    TestDriver() {}
+    static void add_command(TestCommand& cmd);
+    static void add_top_level_commands();
+    static void p4c_cmdl_run_command(TestCommandContext* ctx, TestCommands& commands);
+    static void p4ctest_register(std::string test_name, CommandHandler handler);
+};
+
+}  // namespace Test
+
+/* Overview
+ * ========
+ *
+ * P4C tests directory contains many small test programs. One down side
```

Comment:
I think that this comment should not describe the history of the program creation, it should just say what the program does. This information should be in the pull request, but not in a comment - it will become obsolete after a while.